### PR TITLE
ISSUE-4275, metactl: dump data from a running metasrv

### LINKED ISSUE #4275: Feature: metactl: dump data from a running metasrv
`databend-metactl --export --raft-dir <path>` is used to export meta data for backup when the metasrv is closed.
Running this command with a running metasrv just results in a locking error.

New feature request: `databend-metactl --export --port <port>` to export meta data from a running metasrv on local host.

This can be done by calling the grpc API `export`:
https://github.com/datafuselabs/databend/blob/3713168845ecd80e02d52db386623806a10982d1/metasrv/src/api/grpc/grpc_service.rs#L150-L163

### PR DESCRIPTION
**Pull request by light4**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. `metactl`: dump data from a running `metasrv`
2. fix export panic if data length is smaller than chunk
3. add reflection to grpc service, so we can test it using tools like https://github.com/fullstorydev/grpcurl, will remove it if not needed

## Changelog

- New Feature
- Bug Fix
- Improvement
- Build/Testing/CI
- Need Documentation (Need to add documentation to https://databend.rs)
- Other 

## Related Issues

Fixes #4275

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: tools/metactl/src/main.rs [hunk 1]
```diff
@@ -72,6 +79,14 @@ async fn main() -> anyhow::Result<()> {
 
     eprintln!("raft_config: {}", pretty(raft_config)?);
 
+    // export from grpc api if raft is running
+    let addr = raft_config.raft_api_addr().await?.to_string();
+    if config.export && service_is_running(&addr).await? {
+        eprintln!("export meta from: {}", &config.grpc_api_address);
+        grpc::export_meta(&config.grpc_api_address).await?;
```
Comment (drmingdrmer) [inline tools/metactl/src/main.rs:86]: Why not just check `grpc_api_address` but `raft_api_addr()`?

BTW it exports from a running metasrv if and only if `grpc_api_address` is specified.

File: tools/metactl/src/main.rs [hunk 2]
```diff
@@ -169,3 +184,12 @@ fn print_meta() -> anyhow::Result<()> {
 
     Ok(())
 }
+
+// if port is open, service is running
+async fn service_is_running(addr: &str) -> anyhow::Result<bool> {
```
Comment (drmingdrmer) [inline tools/metactl/src/main.rs:189]: `anyhow::Result` will spread everywhere. Only use it at the top-level entry.

Here a specific error type would be better.

File: tools/metactl/src/grpc.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    let mut lines = vec![];
+    while let Some(chunk_res) = stream.next().await {
+        let chunk = chunk_res?;
+
+        lines.extend_from_slice(&chunk.data);
+    }
+
+    for line in lines {
+        println!("{}", line);
+    }
```
Comment (drmingdrmer) [inline tools/metactl/src/grpc.rs:37]: The lines could be just printed without caching them.

**Review by drmingdrmer — COMMENTED** [review]
Thanks man!

Reflection server is an interesting feature!

There should be a test to ensure this new API works as expected.
Other test about metactl is in `tests/metactl/test-metactl.sh`.

File: tools/metactl/src/grpc.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_meta_grpc::MetaGrpcClient;
+use common_meta_types::protobuf::Empty;
+use tokio_stream::StreamExt;
+
+pub async fn export_meta(addr: &str) -> anyhow::Result<()> {
+    let client = MetaGrpcClient::try_create(addr, "root", "xxx", None, None).await?;
```
Comment (BohuTANG) [inline tools/metactl/src/grpc.rs:20]: password need parameter setting?

File: tools/metactl/src/main.rs (hunk 1 above)
Comment (light4) [reply tools/metactl/src/main.rs:86]: Should use `grpc_api_address` indeed.
What if `metasrv` is running and `grpc_api_address` is not specified.

File: tools/metactl/src/grpc.rs (hunk 4 above)
Comment (light4) [reply tools/metactl/src/grpc.rs:20]: I find this use case in other codes. Wondering it needs an empty user/password or this(root/xxx) if connected from localhost.

File: tools/metactl/src/main.rs (hunk 1 above)
Comment (drmingdrmer) [reply tools/metactl/src/main.rs:86]: If  `grpc_api_address` is invalid metasrv won't run at all.

Besides, if `grpc_api_address` is invalid, there is no way to export data from it.

File: tools/metactl/src/main.rs (hunk 2 above)
Comment (light4) [reply tools/metactl/src/main.rs]: Updated

File: tools/metactl/src/main.rs (hunk 1 above)
Comment (light4) [reply tools/metactl/src/main.rs]: Updated, and a basic test.

File: tests/metactl/test-metactl.sh [hunk 5]
```diff
@@ -16,3 +16,14 @@ cat $meta_json \
...
+# test export from grpc
+chmod +x ./target/debug/databend-meta
+./target/debug/databend-meta --single &
+METASRV_PID=$!
+echo $METASRV_PID
+sleep 0.5
+
+./target/debug/databend-metactl --export
+
+kill $METASRV_PID
```
Comment (drmingdrmer) [inline tests/metactl/test-metactl.sh:29]: The output of `./target/debug/databend-metactl --export` should be redirected to a file and the content of the file should be checked.

Comment (light4) [reply tests/metactl/test-metactl.sh:29]: I wonder if `databend-meta --single` exported data is identical, if so I'll use that.

Comment (drmingdrmer) [reply tests/metactl/test-metactl.sh:29]: I can not tell if it is identical. But there should be at least a record like this:
```
["test-29000-raft_state",{"RaftStateKV":{"key":"Id","value":{"NodeId":0}}}]
```

Comment (light4) [reply tests/metactl/test-metactl.sh]: I find it's not identical. So just check `["raft_state",{"RaftStateKV":{"key":"Id","value":{"NodeId":0}}}]` in exported file.
Wondering if it's ok.

**Review by ariesdevil — APPROVED** [review]

**Pull request merged**